README.md
# megascreen
Atmospheric phase screen generator using a "woofer-tweeter" algorithm


CHANGES.txt
1.0.1
-----
- Updated demos so they don't crash with recent assert
- Demos can run without installing MegaScreen
- Streamlit demo app can run on Streamlit Share

1.0.0
-----
- Added an assert to check that the size of a window does not exceed the size of a tweeter screen
- Removed restriction on dx and increased memory efficiency by allocating the minimum number of tiles needed

0.5.0:
------
- Added a numIter parameter to MegaScreen() to allow straightforward use in for loops
- All parameters to MegaScreen() now have default values to ease adoption
- Tidied up test code
- Added extensive documentation on http://megascreen.readthedocs.io


tests/index.py

# coding: utf-8

# In his blog post [Embedding Matplotlib Animations in IPython Notebooks](http://jakevdp.github.io/blog/2013/05/12/embedding-matplotlib-animations/), Jake VanderPlas presents a slick hack for embedding Matplotlib Animations in IPython Notebooks, which involves writing it as a video to a [tempfile](https://docs.python.org/2/library/tempfile.html), and then re-encoding it in Base64 as a HTML5 Video.
# 
# Unfortunately (or rather fortunately), this hack has been largely rendered obsolete by the heavy development efforts dedicated to both Matplotlib and IPython Notebook ([since renamed to Jupyter Notebook](https://blog.jupyter.org/2015/08/12/first-release-of-jupyter/)) in recent years. In particular, [Matplotlib 1.5.1](http://matplotlib.org/users/whats_new.html#new-in-matplotlib-1-5) now [supports inline display of animations in the notebook](http://matplotlib.org/users/whats_new.html#display-hook-for-animations-in-the-ipython-notebook) with the `to_html5_video` method, which converts the animation to an h264 encoded video and embeddeds it directly in the notebook.
# 
# In this notebook, we reproduce Jake VanderPlas' blog post with this new feature.
# 
# <!-- TEASER_END -->

# In[1]:

get_ipython().magic('matplotlib inline')


# In[2]:

import numpy as np
import matplotlib.pyplot as plt

from matplotlib import animation, rc
from IPython.display import HTML


# In[3]:

# First set up the figure, the axis, and the plot element we want to animate
fig, ax = plt.subplots()

ax.set_xlim(( 0, 2))
ax.set_ylim((-2, 2))

line, = ax.plot([], [], lw=2)


# In[4]:

# initialization function: plot the background of each frame
def init():
    line.set_data([], [])
    return (line,)


# In[5]:

# animation function. This is called sequentially
def animate(i):
    x = np.linspace(0, 2, 1000)
    y = np.sin(2 * np.pi * (x - 0.01 * i))
    line.set_data(x, y)
    return (line,)


# In[6]:

# call the animator. blit=True means only re-draw the parts that have changed.
anim = animation.FuncAnimation(fig, animate, init_func=init,
                               frames=100, interval=20, blit=True)


# In[7]:

HTML(anim.to_html5_video())


# Note that [Animation](http://matplotlib.org/api/animation_api.html#matplotlib.animation.Animation) instances now have a `_repr_html_` method. However, it returns `None` by default.

# In[8]:

anim._repr_html_() is None


# This means we won't get any sort of animation from the inline display.

# In[9]:

anim


# The method used to display is controlled by the `animation.html` rc parameter, which currently supports values of `none` and `html5`. `none` is the default, performing no display. We simply need to set it to `html5`:

# In[10]:

# equivalent to rcParams['animation.html'] = 'html5'
rc('animation', html='html5')


# In[11]:

anim


# And that's all there is to it!


tests/theory.py
"""
Calculate theoretical predictions to compare against simulations
"""
import numpy as np
from numpy import pi
from scipy.special import gamma, jn, j1
import scipy.integrate
from MegaScreen import VonKarmanSpectrum


def winker_variance_quad(n, R=1, r0=2, L0=1e8):
    """Evaluate the variance of a Zernike of radial order n using the method of Winker 1991"""
    return (
        scipy.integrate.quad(
            lambda f: jn(n + 1, (2 * np.pi) * f) ** 2
            * VonKarmanSpectrum(f / R, r0, L0)
            / f,
            0.0,
            np.inf,
        )[0]
        * (n + 1)
        * (2 / np.pi)
        / R ** 2
        * (2 * R / r0) ** (5 / 3)
    )


def jinc_filter(f, fmin):
    w = 2 * pi * f
    return (1 - (4 * (j1(w) / (w)) ** 2)) if f > fmin else (w ** 2 / 4)


def winker_piston_residual(r0=1.0, L0=1e8, lower_limit=0, fmin=1e-4):
    return (
        scipy.integrate.quad(
            lambda f: (
                (1 - (4 * (j1(2 * pi * f) / (2 * pi * f)) ** 2))
                if f > fmin
                else (2 * pi * f) ** 2 / 4
            )
            * f
            * VonKarmanSpectrum(f, r0, L0),
            lower_limit,
            np.inf,
        )[0]
        * (2 * pi)
        * 2 ** (-5 / 3)
    )


def interf_spectrum_quad(baseline, frequencies, r0, L0, eps1=1.0):
    return np.array(
        [
            8
            * (
                scipy.integrate.quad(
                    lambda u, f, b, r0, L0: VonKarmanSpectrum(
                        np.sqrt(u ** 2 + f ** 2), r0, L0
                    )
                    * (1.0 - np.cos(2 * np.pi * u * b)),
                    0,
                    eps1 / baseline,
                    args=(f, baseline, r0, L0),
                    epsrel=1e-3,
                    limit=400,
                )[0]
                + scipy.integrate.quad(
                    lambda u, f, b, r0, L0: VonKarmanSpectrum(
                        np.sqrt(u ** 2 + f ** 2), r0, L0
                    ),
                    eps1 / baseline,
                    np.inf,
                    args=(f, baseline, r0, L0),
                    epsrel=1e-4,
                    limit=100,
                )[0]
                - scipy.integrate.quad(
                    lambda u, f, b, r0, L0: VonKarmanSpectrum(
                        np.sqrt(u ** 2 + f ** 2), r0, L0
                    ),
                    eps1 / baseline,
                    np.inf,
                    args=(f, baseline, r0, L0),
                    epsrel=1e-4,
                    weight="cos",
                    wvar=2 * np.pi * baseline,
                    limit=100,
                )[0]
            )
            for f in frequencies
        ]
    )


def sf_integrated(x, r0=1.0, L0=2e9):
    return [
        2
        * scipy.integrate.quad(
            lambda f, r, r0, L0: VonKarmanSpectrum(f, r0, L0)
            * (1.0 - scipy.special.jn(0, 2 * np.pi * f * r))
            * 2
            * pi
            * f,
            0,
            np.inf,
            args=(r, r0, L0),
            epsrel=1e-3,
        )[0]
        for r in x
    ]


tests/Zernike.py
# Functions to calculate Zernike polynomials
#

import numpy as np
from numpy import sqrt
import functools
import sys


def NumZernike(m):
    """Return the number of polynomials up to and including radial order m"""
    return (m + 1) * (m + 2) // 2


def jtonm(j):
    """Convert Noll index j (1-based) to radial index n and azimuthal index m.  

    The method is described by V. N. Mahajan, Chapter 13, 
    Optical Shop Testing, 3rd. Ed., D. Malacara. ed., Wiley,2007"""
    n = int(sqrt(2 * j - 1) + 0.5) - 1
    m = (1 - 2 * (j % 2)) * (
        2 * int((2 * j + 1 + (n % 2) - n * (n + 1)) / 4.0) - (n % 2)
    )
    return (n, m)


@functools.lru_cache()
def ZernikeGrid(gridSize, maxRadial, diameter=None, orthoganalise=True):
    """Return Zernike polynomials sampled on a square grid.

    Returns all the polynomials for radial orders up to and including maxRadial.
    If diameter is None, then the unit circle just touches the outer edges of the 
    square (with the convention that the square extends +-0.5 pixels beyond the grid).
    Uses a recursive method to evaluate the radial polynomial.
    Returns the polynomials in the order given by Noll (1976), but zero-based.
    """
    if diameter == None:
        diameter = gridSize
    radius = diameter / 2.0
    y, x = np.mgrid[0:gridSize, 0:gridSize]
    x = (x - (gridSize - 1.0) / 2.0) / radius
    y = (y - (gridSize - 1.0) / 2.0) / radius
    # Convert to polar co-ordinates
    temp = x + 1j * y
    r = np.abs(temp)
    R = KinterRadial(r, maxRadial)
    # Create entheta[n] = exp(i*n*theta) recursively from exp(i*theta)
    eitheta = np.where(r == 0.0, 1.0, temp / r)
    entheta = np.concatenate(
        (
            np.ones((1,) + eitheta.shape, dtype=np.complex),
            np.cumprod(np.broadcast_to(eitheta, (maxRadial,) + eitheta.shape), axis=0),
        )
    )
    cntheta = entheta.real
    sntheta = entheta.imag
    jmax = NumZernike(maxRadial)
    Z = np.empty((jmax,) + r.shape)
    for j in range(jmax):
        n, m = jtonm(j + 1)
        const = sqrt((2 - (m == 0)) * (n + 1))
        if m < 0:
            Z[j] = R[n, -m] * sntheta[-m] * const
        else:
            Z[j] = R[n, m] * cntheta[m] * const
    # Make zernike zero outside unit circle (useful for dot product)
    Z = Z * np.less_equal(r, 1.0)
    if orthoganalise:
        return Orthoganalise(Z)
    else:
        return Z


def KinterRadial(r, maxRadial):
    """Compute Zernike radial functions using the modified Kintner method.

    Discussion may be found in Chong, Chee-Way, P. Raveendran, and R. Mukundan. 
    'A Comparative Analysis of Algorithms for Fast Computation of Zernike Moments'.
    Pattern Recognition 36, no. 3 (March 2003): 731-42. 
    doi:10.1016/S0031-3203(02)00091-2."""
    # Storage array for result
    # Creates array 4 * larger than necessary, but simpler to code
    R = np.empty((maxRadial + 1, maxRadial + 1) + r.shape)
    # Calc diagonal elements recursively
    R[0, 0] = np.ones_like(r)
    q = np.arange(1, maxRadial + 1)
    R[q, q] = np.cumprod(np.broadcast_to(r, (maxRadial,) + r.shape), axis=0)
    # Calc second batch of elements according to Fig 3 in Chong et al.
    for q in range(maxRadial - 1):
        R[q + 2, q] = (q + 2) * R[q + 2, q + 2] - (q + 1) * R[q, q]
    # Calc remaining elements
    r2 = r ** 2
    for q in range(maxRadial - 3):
        for p in range(q + 4, maxRadial + 1, 2):
            k1 = (p + q) * (p - q) * (p - 2) / 2.0
            k2 = 2 * p * (p - 1) * (p - 2)
            k3 = -q * q * (p - 1) - p * (p - 1) * (p - 2)
            k4 = -p * (p + q - 2) * (p - q - 2) / 2.0
            R[p, q] = ((k2 / k1) * r2 + (k3 / k1)) * R[p - 2, q] + (k4 / k1) * R[
                p - 4, q
            ]
    return R


def ZernikeHardCoded(gridSize, maxRadial, diameter=None):
    """
    Hardcoded Zernike routine: works only up to 5th radial order
    """
    assert maxRadial <= 5
    if diameter == None:
        diameter = gridSize
    radius = diameter / 2.0
    y, x = np.mgrid[0:gridSize, 0:gridSize]
    x = (x - (gridSize - 1.0) / 2.0) / radius
    y = (y - (gridSize - 1.0) / 2.0) / radius
    # Derive radius and exp(i*theta)
    temp = x + 1j * y
    r1 = np.abs(temp)
    eitheta = np.where(r1 == 0.0, 1.0, temp / r1)
    # Generate powers of r recursively
    r2 = r1 * r1
    r3 = r2 * r1
    r4 = r3 * r1
    r5 = r4 * r1
    # Generate cos and sin terms recursively from exp(i*theta)
    e2 = eitheta * eitheta
    e3 = e2 * eitheta
    e4 = e3 * eitheta
    e5 = e4 * eitheta
    ctheta = eitheta.real
    stheta = eitheta.imag
    c2theta = e2.real
    s2theta = e2.imag
    c3theta = e3.real
    s3theta = e3.imag
    c4theta = e4.real
    s4theta = e4.imag
    c5theta = e5.real
    s5theta = e5.imag
    # Generate all the zernikes
    zernike = np.zeros((21,) + x.shape)
    zernike[0] = 1.0
    zernike[1] = 2.0 * r1 * ctheta
    zernike[2] = 2.0 * r1 * stheta
    zernike[3] = sqrt(3.0) * (2.0 * r2 - 1.0)
    zernike[4] = sqrt(6.0) * r2 * s2theta
    zernike[5] = sqrt(6.0) * r2 * c2theta
    zernike[6] = sqrt(8.0) * (3.0 * r3 - 2.0 * r1) * stheta
    zernike[7] = sqrt(8.0) * (3.0 * r3 - 2.0 * r1) * ctheta
    zernike[8] = sqrt(8.0) * r3 * s3theta
    zernike[9] = sqrt(8.0) * r3 * c3theta
    zernike[10] = sqrt(5.0) * (6.0 * r4 - 6.0 * r2 + 1.0)
    zernike[11] = sqrt(10.0) * (4.0 * r4 - 3.0 * r2) * c2theta
    zernike[12] = sqrt(10.0) * (4.0 * r4 - 3.0 * r2) * s2theta
    zernike[13] = sqrt(10.0) * r4 * c4theta
    zernike[14] = sqrt(10.0) * r4 * s4theta
    zernike[15] = sqrt(12.0) * (10 * r5 - 12 * r3 + 3 * r1) * ctheta
    zernike[16] = sqrt(12.0) * (10 * r5 - 12 * r3 + 3 * r1) * stheta
    zernike[17] = sqrt(12.0) * (5 * r5 - 4 * r3) * c3theta
    zernike[18] = sqrt(12.0) * (5 * r5 - 4 * r3) * s3theta
    zernike[19] = sqrt(12.0) * r5 * c5theta
    zernike[20] = sqrt(12.0) * r5 * s5theta
    # Make zernike zero outside unit circle (useful for dot product)
    zernike = zernike * np.less_equal(r1, 1.0)
    return zernike[: NumZernike(maxRadial)]


def Orthoganalise(Zernikes):
    """Orthoganalise Zernikes to mitigate effects of under-sampling.
    """
    colZernikes = np.transpose(np.reshape(Zernikes, (len(Zernikes), -1)))
    q, r = np.linalg.qr(colZernikes, mode="reduced")
    orthoZernikes = np.reshape(np.transpose(q), Zernikes.shape)
    zernikeNorm = r[0,0]
    orthoZernikes *= zernikeNorm
    return orthoZernikes


def test_vs_hardcoded(gridSize=100, maxRadial=5):
    Z1 = ZernikeGrid(gridSize, maxRadial, orthoganalise=False)
    Z2 = ZernikeHardCoded(gridSize, maxRadial)
    diff = Z1 - Z2
    rms = np.sqrt(np.sum(diff ** 2, axis=(-1, -2)))
    print(rms)
    print(np.amax(rms))




MegaScreen/__init__.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import numpy as np
from numpy import sqrt, fft, random, pi
import functools
import scipy.interpolate

__version__ = "1.0.1"


def FrequencyGrid(shape, pixelSize=1.0):
    """Return a 2-d grid with absolute frequency relevant to an FFT of a
    grid of pixels of size pixelSize."""
    return sqrt(
        np.add.outer(
            np.fft.fftfreq(shape[0], pixelSize) ** 2,
            np.fft.fftfreq(shape[1], pixelSize) ** 2,
        )
    )


def VonKarmanSpectrum(f, r0, L0=1e6, alpha=11.0 / 3.0):
    """Phase spectrum of atmospheric seeing with Von Karman turbulence"""
    return 0.0229 * r0 ** (2.0 - alpha) * (f ** 2 + 1 / L0 ** 2) ** (-alpha / 2)


def FftScreen(spectrum, shape, pixelSize=1.0):
    """Generate infinite sequence of screens based on filtered 2D white noise

    Parameters
    ----------
    spectrum : Callable[[numpy.ndarray[float]],numpy.ndarray[float]]
       spectrum of fluctuations, assumed radially symmetric
    shape: Tuple[int,int]
        Size of output grid
    pixelSize: float
        Pixel size of output grid

    Yields
    ------
    out : ndarray
        2D array of phase disturbances
    """
    f = FrequencyGrid(shape, pixelSize)
    filter = sqrt(spectrum(f) * f[0, 1] * f[1, 0])
    while 1:
        sample = random.normal(size=(2,) + filter.shape)
        result = fft.fft2(filter * (sample[0] + 1j * sample[1]))
        yield result.real
        yield result.imag


def McGlameryScreen(r0, L0=1e5, nfft=256):
    return FftScreen(
        functools.partial(VonKarmanSpectrum, r0=r0, L0=L0), shape=(nfft, nfft)
    )


def SplineTiles(tileGenerator):
    """Generate a sequence of splined tiles with shape (n0/2,n1) from a
    sequence of tiles with shape (n0,n1)
    """
    previous = next(tileGenerator)
    n0 = previous.shape[0] // 2
    assert n0 * 2 == previous.shape[0]
    cspline = np.cos(np.linspace(0, pi / 2, n0, endpoint=False))
    sspline = np.sin(np.linspace(0, pi / 2, n0, endpoint=False))
    for current in tileGenerator:
        yield previous[n0:] * cspline[:, np.newaxis] + current[:n0] * sspline[
            :, np.newaxis
        ]
        previous = current


def GridInterpolator(grid):
    # print(grid.shape)
    xgrid = np.arange(grid.shape[0])
    ygrid = np.arange(grid.shape[1])
    return scipy.interpolate.RectBivariateSpline(xgrid, ygrid, grid)


def SlidingPixels(tileGenerator, x, y, dx):
    """
    Return phase values from a set of pixel coordinates sliding along an infinite ribbon.

    Parameters
    ----------
    tileGenerator: iterator
        sequence of 2D phase screens which are stiched together to form the ribbon
    x,y: 1D arrays
        relative pixel coordinates in units of the tile grid size
    dx: float
        increment of pixel "x" coordinate on each iteration

    Yields
    -------
    1D array
        phase values at each pixel for this iteration
    """
    # Shift the x and y coordinates so that the minimum coordinate is zero
    x_zeroed = x - np.amin(x)
    y_zeroed = y - np.amin(y)
    xmax = np.amax(x_zeroed)
    ymax = np.amax(y_zeroed)
    tiles = [next(tileGenerator)]
    xtile, ytile = tiles[0].shape
    assert ymax <= ytile  # No limit on xmax, except memory to hold ribbon
    # Fill up initial ribbon
    for i in range(int(np.ceil(xmax / xtile))):
        tiles.append(next(tileGenerator))
    re_interpolate = True
    xoffset = 0
    while True:
        if re_interpolate:
            interpolator = GridInterpolator(np.concatenate(tiles))
            re_interpolate = False
        yield interpolator(x_zeroed + xoffset, y_zeroed, grid=False)
        xoffset += dx
        while xoffset > xtile:
            tiles.pop(0)
            tiles.append(next(tileGenerator))
            re_interpolate = True
            xoffset -= xtile


def PixelCoords(origin, shape, pixelSize=1, theta=0):
    """Return x and y coodinates of a grid of pixels in rectangular region
    given by *origin* and *shape*, in a frame scaled to *pixelSize* and
    rotated by angle *theta*
    """
    c = np.cos(theta)
    s = np.sin(theta)
    # print(origin, shape, pixelSize)
    x = (origin[0] + np.arange(shape[0])) * pixelSize
    y = (origin[1] + np.arange(shape[1])) * pixelSize
    return np.add.outer(c * x, s * y).flatten(), np.add.outer(-s * x, c * y).flatten()


def SlidingWindows(
    tileGenerator, shape, dx, origins=((0.0, 0.0),), pixelSize=1, theta=0.0
):
    """Return phase values from a set of rectangular windows sliding along an infinite ribbon.

    Parameters
    ----------
    tileGenerator: iterator
        A sequence of 2D phase screens which are stiched together to form a ribbon
    origins: sequence of pairs of floats
        The origins of each of the rectangular windows
    shape: tuple
        Shape of rectangular window (same for all windows)
    dx: float
        Increment of pixel `x` coordinate on each iteration
     Yields
     -------
     2D array
         phase values at each pixel
     """

    coords = [
        PixelCoords(origin=origin, shape=shape, pixelSize=pixelSize, theta=theta)
        for origin in origins
    ]
    # print(coords)
    coords = np.array(coords)
    x = coords[:, 0, :].flat
    y = coords[:, 1, :].flat
    numWindow = len(origins)
    if numWindow == 1:
        newshape = shape
    else:
        newshape = [numWindow] + list(shape)
    for screen in SlidingPixels(tileGenerator, x, y, dx):
        yield np.reshape(screen, newshape)


def NestedSpectra(spectrum, f0, eps=1e-6):
    grad = (spectrum(f0 * (1 + eps)) - spectrum(f0 * (1 - eps))) / (2 * f0 * eps)
    c1 = spectrum(f0)

    def OuterSpectrum(f):
        s = spectrum(f)
        s1 = np.where(
            f < f0, c1 - 2 * grad * f0 / np.pi * np.cos(np.pi * f / (2 * f0)), s
        )
        return np.where(s1 < s, s1, s)

    def InnerSpectrum(f):
        return spectrum(f) - OuterSpectrum(f)

    return InnerSpectrum, OuterSpectrum


def NestedScreen(
    spectrum,
    windowShape,
    dx,
    windowOrigins=((0.0, 0.0),),
    pixelSize=1.0,
    theta=0.0,
    nfftWoofer=256,
    nfftTweeter=256,
    frequencyOverlap=4.0,
    fractionalSupport=0.5,
    debug=False,
    numIter=None,
):
    """Generate a sequence of phase screens for an arbitrary spectrum

    Parameters
    ----------
    spectrum: Callable[[numpy.ndarray[float]],numpy.ndarray[float]]
       Returns the spectral power density of the phase perturbations at a given frequency
    Notes
    -----
    See MegaScreen() for the other parameters
    """
    wooferPixelSize = nfftTweeter / (2 * frequencyOverlap)
    f0 = 1 / (2 * wooferPixelSize) * fractionalSupport
    wooferSpectrum, tweeterSpectrum = NestedSpectra(spectrum, f0)
    innerWindows = SlidingWindows(
        SplineTiles(
            FftScreen(wooferSpectrum, (nfftWoofer, nfftWoofer), wooferPixelSize)
        ),
        dx=dx / wooferPixelSize,
        shape=windowShape,
        origins=windowOrigins,
        pixelSize=pixelSize / wooferPixelSize,
        theta=theta,
    )
    outerWindows = [
        SlidingWindows(
            SplineTiles(
                FftScreen(tweeterSpectrum, (nfftTweeter, nfftTweeter), pixelSize)
            ),
            dx=dx,
            shape=windowShape,
            origins=[origin],
            pixelSize=pixelSize,
            theta=theta,
        )
        for origin in windowOrigins
    ]
    iter = 0
    while numIter is None or iter < numIter:
        iter += 1
        inner = next(innerWindows)
        outer = np.squeeze(np.array([next(o) for o in outerWindows]))
        if debug:
            yield inner, outer, inner + outer
        else:
            yield inner + outer


def MegaScreen(
    r0=7.0,
    L0=7000.0,
    windowShape=(100, 100),
    dx=3.5,
    windowOrigins=((0.0, 0.0),),
    pixelSize=1.0,
    theta=0.0,
    nfftWoofer=256,
    nfftTweeter=256,
    frequencyOverlap=4.0,
    fractionalSupport=1.0,
    debug=False,
    numIter=None,
):
    """
    Generate a sequence of phase screens with a Von Karman spectrum.

    Parameters
    ----------

    r0 : float
         Fried parameter :math:`r_0` in tweeter pixel units.
    L0 : float
         Outer scale of turbulence in tweeter pixel units.
    windowShape : Tuple[int,int]
                  Shape of rectangular output window grid (same for all windows).
    dx : float
         Increment in the "x" coordinate of the tweeter phase screen between 
         subsequent calls. Represents the "frozen turbulence" windspeed in 
         tweeter pixels/iteration. Should be > 0. See note below about coordinate 
         directions.
    windowOrigins : Sequence[Tuple[float,float]]
                    Relative coordinates of the rectangular windows in the window 
                    coordinate system - note that this coordinate system is scaled
                    and rotated with respect to the to the coordinate system of 
                    the "woofer" and "tweeter" screens, and hence to the "wind" direction.
    pixelSize : float
                Size of the window pixels in tweeter pixel units 
                (typically <= 1.0).
    theta: float
           Angle in radians between the output window "x" axis and the 
           tweeter screen "x" axis. Used to simulate the wind travelling in a given 
           direction with respect to the window coordinate axes. See note below about
           the coordinate convention used.
    nfftWoofer : int
                 Size of the square FFT used to produce the woofer screen.
    nfftTweeter : int
                 Size of the square FFT used to produce the tweeter screen.
    frequencyOverlap : float
                       The Nyquist frequency of the woofer spectrum in units of the 
                       fundamental frequency of the tweeter spectrum.
    fractionalSupport : float
                        Frequency above which woofer spectrum is zero (the "crossover
                        frequency"), expressed as a fraction of the woofer Nyquist 
                        frequency.
    debug : boolean
            If true, yield additional debugging information along with phase screens.
    numIter : Optional[int]
            Number of iterations to stop after, or None to return an infinite,
            non-repeating sequence of phase screens.

    Yields
    ------

    screen  : numpy.ndarray[float]
              Wavefront perturbation at each pixel in each of the output windows, in
              radians. If there is only one window this is a 2-D array, otherwise 
              an array of 2-D arrays (i.e. a 3-D array) is returned.

    Notes
    -----
    The convention used in the above descriptions has the "x" coordinate corresponding 
    to the leftmost index of the 2-D phase screen arrays. This is a FORTRAN-like 
    convention, and when the phase screen is plotted in `matplotlib.imshow()` and similar
    image plotting functions, this coordinate appears as the "y" coordinate in the 
    image (albeit by default the "y" coordinate is plotted increasing downwards).
    """
    spectrum = functools.partial(VonKarmanSpectrum, r0=r0, L0=L0)
    return NestedScreen(
        spectrum,
        windowShape,
        dx,
        windowOrigins=windowOrigins,
        pixelSize=pixelSize,
        theta=theta,
        nfftWoofer=nfftWoofer,
        nfftTweeter=nfftTweeter,
        frequencyOverlap=frequencyOverlap,
        fractionalSupport=fractionalSupport,
        debug=debug,
        numIter=numIter,
    )


MegaScreen/Zernike.py
# Functions to calculate Zernike polynomials
#

import numpy as np
from numpy import sqrt
import functools
import sys


def NumZernike(m):
    """Return the number of polynomials up to and including radial order m"""
    return (m + 1) * (m + 2) // 2


def jtonm(j):
    """Convert Noll index j (1-based) to radial index n and azimuthal index m.  

    The method is described by V. N. Mahajan, Chapter 13, 
    Optical Shop Testing, 3rd. Ed., D. Malacara. ed., Wiley,2007"""
    n = int(sqrt(2 * j - 1) + 0.5) - 1
    m = (1 - 2 * (j % 2)) * (
        2 * int((2 * j + 1 + (n % 2) - n * (n + 1)) / 4.0) - (n % 2)
    )
    return (n, m)


@functools.lru_cache()
def ZernikeGrid(gridSize, maxRadial, diameter=None, orthoganalise=True):
    """Return Zernike polynomials sampled on a square grid.

    Returns all the polynomials for radial orders up to and including maxRadial.
    If diameter is None, then the unit circle just touches the outer edges of the 
    square (with the convention that the square extends +-0.5 pixels beyond the grid).
    Uses a recursive method to evaluate the radial polynomial.
    Returns the polynomials in the order given by Noll (1976), but zero-based.
    """
    if diameter == None:
        diameter = gridSize
    radius = diameter / 2.0
    y, x = np.mgrid[0:gridSize, 0:gridSize]
    x = (x - (gridSize - 1.0) / 2.0) / radius
    y = (y - (gridSize - 1.0) / 2.0) / radius
    # Convert to polar co-ordinates
    temp = x + 1j * y
    r = np.abs(temp)
    R = KinterRadial(r, maxRadial)
    # Create entheta[n] = exp(i*n*theta) recursively from exp(i*theta)
    eitheta = np.where(r == 0.0, 1.0, temp / r)
    entheta = np.concatenate(
        (
            np.ones((1,) + eitheta.shape, dtype=np.complex),
            np.cumprod(np.broadcast_to(eitheta, (maxRadial,) + eitheta.shape), axis=0),
        )
    )
    cntheta = entheta.real
    sntheta = entheta.imag
    jmax = NumZernike(maxRadial)
    Z = np.empty((jmax,) + r.shape)
    for j in range(jmax):
        n, m = jtonm(j + 1)
        const = sqrt((2 - (m == 0)) * (n + 1))
        if m < 0:
            Z[j] = R[n, -m] * sntheta[-m] * const
        else:
            Z[j] = R[n, m] * cntheta[m] * const
    # Make zernike zero outside unit circle (useful for dot product)
    Z = Z * np.less_equal(r, 1.0)
    if orthoganalise:
        return Orthoganalise(Z)
    else:
        return Z


def KinterRadial(r, maxRadial):
    """Compute Zernike radial functions using the modified Kintner method.

    Discussion may be found in Chong, Chee-Way, P. Raveendran, and R. Mukundan. 
    'A Comparative Analysis of Algorithms for Fast Computation of Zernike Moments'.
    Pattern Recognition 36, no. 3 (March 2003): 731-42. 
    doi:10.1016/S0031-3203(02)00091-2."""
    # Storage array for result
    # Creates array 4 * larger than necessary, but simpler to code
    R = np.empty((maxRadial + 1, maxRadial + 1) + r.shape)
    # Calc diagonal elements recursively
    R[0, 0] = np.ones_like(r)
    q = np.arange(1, maxRadial + 1)
    R[q, q] = np.cumprod(np.broadcast_to(r, (maxRadial,) + r.shape), axis=0)
    # Calc second batch of elements according to Fig 3 in Chong et al.
    for q in range(maxRadial - 1):
        R[q + 2, q] = (q + 2) * R[q + 2, q + 2] - (q + 1) * R[q, q]
    # Calc remaining elements
    r2 = r ** 2
    for q in range(maxRadial - 3):
        for p in range(q + 4, maxRadial + 1, 2):
            k1 = (p + q) * (p - q) * (p - 2) / 2.0
            k2 = 2 * p * (p - 1) * (p - 2)
            k3 = -q * q * (p - 1) - p * (p - 1) * (p - 2)
            k4 = -p * (p + q - 2) * (p - q - 2) / 2.0
            R[p, q] = ((k2 / k1) * r2 + (k3 / k1)) * R[p - 2, q] + (k4 / k1) * R[
                p - 4, q
            ]
    return R


def ZernikeHardCoded(gridSize, maxRadial, diameter=None):
    """
    Hardcoded Zernike routine: works only up to 5th radial order
    """
    assert maxRadial <= 5
    if diameter == None:
        diameter = gridSize
    radius = diameter / 2.0
    y, x = np.mgrid[0:gridSize, 0:gridSize]
    x = (x - (gridSize - 1.0) / 2.0) / radius
    y = (y - (gridSize - 1.0) / 2.0) / radius
    # Derive radius and exp(i*theta)
    temp = x + 1j * y
    r1 = np.abs(temp)
    eitheta = np.where(r1 == 0.0, 1.0, temp / r1)
    # Generate powers of r recursively
    r2 = r1 * r1
    r3 = r2 * r1
    r4 = r3 * r1
    r5 = r4 * r1
    # Generate cos and sin terms recursively from exp(i*theta)
    e2 = eitheta * eitheta
    e3 = e2 * eitheta
    e4 = e3 * eitheta
    e5 = e4 * eitheta
    ctheta = eitheta.real
    stheta = eitheta.imag
    c2theta = e2.real
    s2theta = e2.imag
    c3theta = e3.real
    s3theta = e3.imag
    c4theta = e4.real
    s4theta = e4.imag
    c5theta = e5.real
    s5theta = e5.imag
    # Generate all the zernikes
    zernike = np.zeros((21,) + x.shape)
    zernike[0] = 1.0
    zernike[1] = 2.0 * r1 * ctheta
    zernike[2] = 2.0 * r1 * stheta
    zernike[3] = sqrt(3.0) * (2.0 * r2 - 1.0)
    zernike[4] = sqrt(6.0) * r2 * s2theta
    zernike[5] = sqrt(6.0) * r2 * c2theta
    zernike[6] = sqrt(8.0) * (3.0 * r3 - 2.0 * r1) * stheta
    zernike[7] = sqrt(8.0) * (3.0 * r3 - 2.0 * r1) * ctheta
    zernike[8] = sqrt(8.0) * r3 * s3theta
    zernike[9] = sqrt(8.0) * r3 * c3theta
    zernike[10] = sqrt(5.0) * (6.0 * r4 - 6.0 * r2 + 1.0)
    zernike[11] = sqrt(10.0) * (4.0 * r4 - 3.0 * r2) * c2theta
    zernike[12] = sqrt(10.0) * (4.0 * r4 - 3.0 * r2) * s2theta
    zernike[13] = sqrt(10.0) * r4 * c4theta
    zernike[14] = sqrt(10.0) * r4 * s4theta
    zernike[15] = sqrt(12.0) * (10 * r5 - 12 * r3 + 3 * r1) * ctheta
    zernike[16] = sqrt(12.0) * (10 * r5 - 12 * r3 + 3 * r1) * stheta
    zernike[17] = sqrt(12.0) * (5 * r5 - 4 * r3) * c3theta
    zernike[18] = sqrt(12.0) * (5 * r5 - 4 * r3) * s3theta
    zernike[19] = sqrt(12.0) * r5 * c5theta
    zernike[20] = sqrt(12.0) * r5 * s5theta
    # Make zernike zero outside unit circle (useful for dot product)
    zernike = zernike * np.less_equal(r1, 1.0)
    return zernike[: NumZernike(maxRadial)]


def Orthoganalise(Zernikes):
    """Orthoganalise Zernikes to mitigate effects of under-sampling.
    """
    colZernikes = np.transpose(np.reshape(Zernikes, (len(Zernikes), -1)))
    q, r = np.linalg.qr(colZernikes, mode="reduced")
    orthoZernikes = np.reshape(np.transpose(q), Zernikes.shape)
    zernikeNorm = np.sqrt(np.sum(r ** 2, axis=0)) * np.sign(np.diag(r))
    orthoZernikes *= zernikeNorm[:, None, None]
    return orthoZernikes


def test_vs_hardcoded(gridSize=100, maxRadial=5):
    Z1 = ZernikeGrid(gridSize, maxRadial, orthoganalise=False)
    Z2 = ZernikeHardCoded(gridSize, maxRadial)
    diff = Z1 - Z2
    rms = np.sqrt(np.sum(diff ** 2, axis=(-1, -2)))
    print(rms)
    print(np.amax(rms))


def test_ortho(gridSize=32, maxRadial=5):
    Z = ZernikeGrid(gridSize, maxRadial)
    orth, q, r = Orthoganalise(Z, debug=True)
    print(q.shape, r.shape, orth.shape)
    print((np.abs(r) > 1e-5).astype(np.int))
    m = gridSize // 2
    print(orth[:3, m - 2 : m + 2, m - 2 : m + 2])
    diag = np.diag(r)
    roff = r - np.diag(diag)
    print(np.amax(roff))
    print(diag ** 2)
    print(np.sum(Z ** 2, axis=(-1, -2)) - np.sum(orth ** 2, axis=(-2, -1)))
    print(np.sum(q ** 2, axis=0))
    # print(np.sum(roff**2,axis=0))
    # print(np.sum(r**2,axis=1))
    # print(np.sum(r**2,axis=0))


test_vs_hardcoded()
# test_ortho()
# Z=ZernikeGrid(gridSize=100, maxRadial=5)
